pkg/operation/botanist/etcd.go

```diff
@@ -191,6 +192,63 @@ func (b *Botanist) scaleETCD(ctx context.Context, replicas int32) error {
 	return b.Shoot.Components.ControlPlane.EtcdEvents.Scale(ctx, replicas)
 }
 
+func (b *Botanist) deployOrRestoreMainEtcd(ctx context.Context) error {
+	isRestoreRequired, err := b.isRestorationOfMultiNodeMainEtcdRequired(ctx)
+	if err != nil {
+		return err
+	}
+
+	if isRestoreRequired {
+		return b.restoreMultiNodeMainEtcd(ctx)
+	}
+
+	return b.Shoot.Components.ControlPlane.EtcdMain.Deploy(ctx)
+}
+
+func (b *Botanist) isRestorationOfMultiNodeMainEtcdRequired(ctx context.Context) (bool, error) {
+	if !b.isRestorePhase() || !v1beta1helper.IsHAControlPlaneConfigured(b.Shoot.GetInfo()) {
+		return false, nil
+	}
+
+	etcd, err := b.Shoot.Components.ControlPlane.EtcdMain.Get(ctx)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+
+	// The etcd has already been scaled up to the desired number of replicas
+	// and therefore has been restored, then it has been scaled down to 0
+	// replicas as part of the reconciliation flow for hibernated shoots.
+	if b.Shoot.HibernationEnabled && etcd.Spec.Replicas == 0 {
+		return false, nil
+	}
+	// The etcd has already been scaled up to the desired number of replicas
+	// and therefore has been restored.
+	if etcd.Spec.Replicas == getEtcdReplicas(b.Shoot.GetInfo()) {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (b *Botanist) restoreMultiNodeMainEtcd(ctx context.Context) error {
+	desiredReplicas := b.Shoot.Components.ControlPlane.EtcdMain.GetReplicas()
+	defer func() {
+		b.Shoot.Components.ControlPlane.EtcdMain.SetReplicas(desiredReplicas)
+	}()
+
+	b.Shoot.Components.ControlPlane.EtcdMain.SetReplicas(pointer.Int32(1))
+	if err := b.Shoot.Components.ControlPlane.EtcdMain.Deploy(ctx); err != nil {
+		return err
+	}
+	if err := b.Shoot.Components.ControlPlane.EtcdMain.Wait(ctx); err != nil {
+		return err
+	}
+	return b.Shoot.Components.ControlPlane.EtcdMain.Scale(ctx, getEtcdReplicas(b.Shoot.GetInfo()))
```

Comment:
This second wait is automatically handled by https://github.com/gardener/gardener/blob/d51027ce9190741620efe83b29489171d9c0890b/pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go#L290-L294

Reply:
As discussed via chat, the comment was written under the assumption that a final last deployment will happen after the `defer` statement which is however not necessary because a potential scale down happens through `HibernateControlPlane` in the shoot control flow.